src/search/tasks/default_value_axioms_task.cc

```diff
@@ -0,0 +1,426 @@
+    // We don't need the sccs if we set axioms "v=default <- {}" everywhere.
+    if (axioms == AxiomHandlingType::APPROXIMATE_NEGATIVE_CYCLES) {
+        sccs = sccs::compute_maximal_sccs(nondefault_dependencies);
+        var_to_scc = vector<vector<int> *>(
+            task_proxy.get_variables().size(), nullptr);
+        for (int i = 0; i < (int)sccs.size(); ++i) {
+            for (int var: sccs[i]) {
+                var_to_scc[var] = &sccs[i];
+            }
+        }
+    }
+
+    unordered_set<int> default_value_needed = get_default_value_needed(
+        nondefault_dependencies, default_dependencies, var_to_scc);
+
+    for (int var: default_value_needed) {
+        vector<int> &axiom_ids = axiom_ids_for_var[var];
+        int default_value =
+            task_proxy.get_variables()[var].get_default_axiom_value();
+
+        if (axioms == AxiomHandlingType::APPROXIMATE_NEGATIVE
+            || var_to_scc[var]->size() > 1) {
+            /*
+               If there is a cyclic dependency between several derived
+               variables, the "obvious" way of negating the formula
+               defining the derived variable is semantically wrong
+               (see issue453).
+
+               In this case we perform a naive overapproximation
+               instead, which assumes that derived variables occurring
+               in the cycle can be false unconditionally. This is good
+               enough for correctness of the code that uses these
+               default value axioms, but loses accuracy. Negating the
+               axioms in an exact (non-overapproximating) way is possible
+               but more expensive (again, see issue453).
+            */
+            default_value_axioms.emplace_back(
+                FactPair(var, default_value), vector<FactPair>());
+        } else {
+            add_default_value_axioms_for_var(
+                FactPair(var, default_value), axiom_ids);
+        }
+    }
+}
+
+/*
+  Collect for which derived variables it is relevant to know how they
+  can obtain their default value. This is done by tracking for all
+  derived variables which of their value are needed.
+
+  Initially, we know that var=val is needed if it appears in a goal or
+  operator condition. Then we iteratively do the following:
+  a) If var=val is needed and var'=nondefault is in the body of an
+     axiom setting var=nondefault, then var'=val is needed.
+  b) If var=val is needed and var'=default is in the body of an axiom
+     setting var=nondefault, then var'=!val is needed, where
+       - !val=nondefault if val=default
+       - !val=default if val=nondefault
+  (var and var' are always derived variables.)
+
+  If var=default is needed but we already know that the axioms we will
+  introduce for var=default are going to have an empty body, then we don't
+  apply a)/b) (because the axiom for var=default will not depend on anything).
+*/
+unordered_set<int> DefaultValueAxiomsTask::get_default_value_needed(
```

Comment:
I am not totally convinced by the name of the method but struggle to come up with something better.
`determine_variables_with_relevant_default_value` is maybe a bit long.

Reply:
I used the slightly shorter `get_vars_with_relevant_default_value`.

Reply:
Sounds good!